File: tests/qos/test_qos_loading.py
"""Step 1.3 Test Gate — Validate QoS XML loading via Python default QosProvider.

Verifies that all QoS XML files load correctly via NDDS_QOS_PROFILES,
snippet/pattern/topic profiles resolve, and topic-filter QoS resolution
returns the expected policy values.

Python API methods (confirmed via rti-chatbot-mcp, INC-003):
    Named profile:      datawriter_qos_from_profile(profile)
    Topic + profile:    set_topic_datawriter_qos(profile, topic_name)
    Topic + default:    get_topic_datawriter_qos(topic_name)
    Participant:        participant_qos_from_profile(profile)
"""

import os

import pytest
import rti.connextdds as dds


@pytest.fixture(autouse=True)
def _check_env():
    """Ensure NDDS_QOS_PROFILES is set before any test runs."""
    assert os.environ.get(
        "NDDS_QOS_PROFILES"
    ), "NDDS_QOS_PROFILES not set — source install/setup.bash first"


@pytest.fixture
def provider():
    """Return the default QosProvider (loads from NDDS_QOS_PROFILES)."""
    return dds.QosProvider.default


# --- Snippets ---


class TestSnippets:
    def test_reliable(self, provider):
        qos = provider.datawriter_qos_from_profile(
            "BuiltinQosSnippetLib::QosPolicy.Reliability.Reliable"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE

    def test_best_effort(self, provider):
        qos = provider.datawriter_qos_from_profile(
            "BuiltinQosSnippetLib::QosPolicy.Reliability.BestEffort"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT

    def test_transient_local(self, provider):
        qos = provider.datawriter_qos_from_profile(
            "BuiltinQosSnippetLib::QosPolicy.Durability.TransientLocal"
        )
        assert qos.durability.kind == dds.DurabilityKind.TRANSIENT_LOCAL

    def test_volatile(self, provider):
        qos = provider.datawriter_qos_from_profile("Snippets::Volatile")
        assert qos.durability.kind == dds.DurabilityKind.VOLATILE

    def test_keep_last_1(self, provider):
        qos = provider.datawriter_qos_from_profile(
            "BuiltinQosSnippetLib::QosPolicy.History.KeepLast_1"
        )
        assert qos.history.kind == dds.HistoryKind.KEEP_LAST
        assert qos.history.depth == 1

    def test_keep_all(self, provider):
        qos = provider.datawriter_qos_from_profile(
            "BuiltinQosSnippetLib::QosPolicy.History.KeepAll"
        )
        assert qos.history.kind == dds.HistoryKind.KEEP_ALL


# --- Patterns ---


class TestPatterns:
    def test_state_writer(self, provider):
        qos = provider.datawriter_qos_from_profile("Patterns::State")
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE
        assert qos.durability.kind == dds.DurabilityKind.TRANSIENT_LOCAL
        assert qos.history.depth == 1

    def test_command_writer(self, provider):
        qos = provider.datawriter_qos_from_profile("Patterns::Command")
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE
        assert qos.durability.kind == dds.DurabilityKind.VOLATILE

    def test_stream_writer(self, provider):
        qos = provider.datawriter_qos_from_profile("Patterns::Stream")
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT
        assert qos.history.depth == 1

    def test_stream_reader(self, provider):
        qos = provider.datareader_qos_from_profile("Patterns::Stream")
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT
        assert qos.history.depth == 4


# --- Topic-filter QoS resolution (named profile) ---


class TestTopicFilters:
    def test_patient_vitals_writer(self, provider):
        """PatientVitals: State base + DeadlinePatientVitals override."""
        qos = provider.set_topic_datawriter_qos(
            "Topics::ProcedureTopics", "PatientVitals"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE
        assert qos.durability.kind == dds.DurabilityKind.TRANSIENT_LOCAL
        assert qos.deadline.period.sec == 2

    def test_patient_vitals_reader(self, provider):
        qos = provider.set_topic_datareader_qos(
            "Topics::ProcedureTopics", "PatientVitals"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE
        assert qos.deadline.period.sec == 2

    def test_operator_input_writer(self, provider):
        """OperatorInput: BestEffort + KeepLast1 (writer) + DeadlineOperatorInput + LifespanOperatorInput."""
        qos = provider.set_topic_datawriter_qos(
            "Topics::ProcedureTopics", "OperatorInput"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT
        assert qos.history.depth == 1
        assert qos.deadline.period.nanosec == 4000000

    def test_robot_command_writer(self, provider):
        """RobotCommand: Command pattern → Reliable + Volatile."""
        qos = provider.set_topic_datawriter_qos(
            "Topics::ProcedureTopics", "RobotCommand"
        )
        assert qos.durability.kind == dds.DurabilityKind.VOLATILE

    def test_robot_state_writer(self, provider):
        """RobotState: State base + DeadlineRobotState."""
        qos = provider.set_topic_datawriter_qos("Topics::ProcedureTopics", "RobotState")
        assert qos.reliability.kind == dds.ReliabilityKind.RELIABLE
        assert qos.deadline.period.nanosec == 20000000

    def test_camera_frame_writer(self, provider):
        """CameraFrame: Stream + DeadlineCameraFrame."""
        qos = provider.set_topic_datawriter_qos(
            "Topics::ProcedureTopics", "CameraFrame"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT
        assert qos.deadline.period.nanosec == 66000000

    def test_waveform_data_writer(self, provider):
        """WaveformData: Stream + DeadlineWaveform."""
        qos = provider.set_topic_datawriter_qos(
            "Topics::ProcedureTopics", "WaveformData"
        )
        assert qos.reliability.kind == dds.ReliabilityKind.BEST_EFFORT
        assert qos.deadline.period.nanosec == 40000000


# --- Participant profiles ---


class TestParticipants:
    def test_transport_loads(self, provider):
        qos = provider.participant_qos_from_profile("Participants::Transport")
        assert qos is not None

    def test_factory_defaults_loads(self, provider):
        qos = provider.participant_qos_from_profile("Factory::FactoryDefaults")
        assert qos is not None

File: tests/qos/test_qos_loading.cpp
/**
 * @file test_qos_loading.cpp
 * @brief Step 1.3 Test Gate — Validate QoS XML loading via default QosProvider.
 *
 * Verifies that all QoS XML files load correctly via NDDS_QOS_PROFILES,
 * snippet/pattern/topic profiles resolve, and topic-filter QoS resolution
 * returns the expected policy values.
 *
 * RTI Extension API (INC-003):
 *   provider->datawriter_qos_w_topic_name(profile, topic_name)
 *   provider->datareader_qos_w_topic_name(profile, topic_name)
 */

#include <gtest/gtest.h>
#include <dds/core/QosProvider.hpp>
#include <dds/pub/qos/DataWriterQos.hpp>
#include <dds/sub/qos/DataReaderQos.hpp>

class QosLoadingTest : public ::testing::Test {
protected:
    dds::core::QosProvider provider_{dds::core::QosProvider::Default()};
};

// --- Snippets: isolated policy chunks ---

TEST_F(QosLoadingTest, SnippetReliable) {
    auto qos = provider_.datawriter_qos("BuiltinQosSnippetLib::QosPolicy.Reliability.Reliable");
    auto kind = qos.policy<dds::core::policy::Reliability>().kind();
    EXPECT_EQ(kind, dds::core::policy::ReliabilityKind::RELIABLE);
}

TEST_F(QosLoadingTest, SnippetBestEffort) {
    auto qos = provider_.datawriter_qos("BuiltinQosSnippetLib::QosPolicy.Reliability.BestEffort");
    auto kind = qos.policy<dds::core::policy::Reliability>().kind();
    EXPECT_EQ(kind, dds::core::policy::ReliabilityKind::BEST_EFFORT);
}

TEST_F(QosLoadingTest, SnippetTransientLocal) {
    auto qos = provider_.datawriter_qos("BuiltinQosSnippetLib::QosPolicy.Durability.TransientLocal");
    auto kind = qos.policy<dds::core::policy::Durability>().kind();
    EXPECT_EQ(kind, dds::core::policy::DurabilityKind::TRANSIENT_LOCAL);
}

TEST_F(QosLoadingTest, SnippetVolatile) {
    auto qos = provider_.datawriter_qos("Snippets::Volatile");
    auto kind = qos.policy<dds::core::policy::Durability>().kind();
    EXPECT_EQ(kind, dds::core::policy::DurabilityKind::VOLATILE);
}

TEST_F(QosLoadingTest, SnippetKeepLast1) {
    auto qos = provider_.datawriter_qos("BuiltinQosSnippetLib::QosPolicy.History.KeepLast_1");
    auto hist = qos.policy<dds::core::policy::History>();
    EXPECT_EQ(hist.kind(), dds::core::policy::HistoryKind::KEEP_LAST);
    EXPECT_EQ(hist.depth(), 1);
}

TEST_F(QosLoadingTest, SnippetKeepAll) {
    auto qos = provider_.datawriter_qos("BuiltinQosSnippetLib::QosPolicy.History.KeepAll");
    auto kind = qos.policy<dds::core::policy::History>().kind();
    EXPECT_EQ(kind, dds::core::policy::HistoryKind::KEEP_ALL);
}

// --- Patterns: composite base profiles ---

TEST_F(QosLoadingTest, PatternStateWriter) {
    auto qos = provider_.datawriter_qos("Patterns::State");
    auto rel = qos.policy<dds::core::policy::Reliability>().kind();
    auto dur = qos.policy<dds::core::policy::Durability>().kind();
    auto depth = qos.policy<dds::core::policy::History>().depth();
    EXPECT_EQ(rel, dds::core::policy::ReliabilityKind::RELIABLE);
    EXPECT_EQ(dur, dds::core::policy::DurabilityKind::TRANSIENT_LOCAL);
    EXPECT_EQ(depth, 1);
}

TEST_F(QosLoadingTest, PatternCommandWriter) {
    auto qos = provider_.datawriter_qos("Patterns::Command");
    auto rel = qos.policy<dds::core::policy::Reliability>().kind();
    auto dur = qos.policy<dds::core::policy::Durability>().kind();
    EXPECT_EQ(rel, dds::core::policy::ReliabilityKind::RELIABLE);
    EXPECT_EQ(dur, dds::core::policy::DurabilityKind::VOLATILE);
}

TEST_F(QosLoadingTest, PatternStreamWriter) {
    auto qos = provider_.datawriter_qos("Patterns::Stream");
    auto rel = qos.policy<dds::core::policy::Reliability>().kind();
    auto depth = qos.policy<dds::core::policy::History>().depth();
    EXPECT_EQ(rel, dds::core::policy::ReliabilityKind::BEST_EFFORT);
    EXPECT_EQ(depth, 1);
}

// --- Topics: topic-filter-bound QoS resolution (RTI extension API) ---

TEST_F(QosLoadingTest, TopicFilterPatientVitals) {
    auto qos = provider_->datawriter_qos_w_topic_name(
        "Topics::ProcedureTopics", "PatientVitals");
    auto rel = qos.policy<dds::core::policy::Reliability>().kind();
    auto dur = qos.policy<dds::core::policy::Durability>().kind();
    auto deadline_sec = qos.policy<dds::core::policy::Deadline>().period().sec();
    EXPECT_EQ(rel, dds::core::policy::ReliabilityKind::RELIABLE);
    EXPECT_EQ(dur, dds::core::policy::DurabilityKind::TRANSIENT_LOCAL);
    EXPECT_EQ(deadline_sec, 2);
}

TEST_F(QosLoadingTest, TopicFilterOperatorInput) {
    auto qos = provider_->datawriter_qos_w_topic_name(
        "Topics::ProcedureTopics", "OperatorInput");
    auto rel = qos.policy<dds::core::policy::Reliability>().kind();
    auto depth = qos.policy<dds::core::policy::History>().depth();
    auto deadline_ns = qos.policy<dds::core::policy::Deadline>().period().nanosec();
    EXPECT_EQ(rel, dds::core::policy::ReliabilityKind::BEST_EFFORT);
    EXPECT_EQ(depth, 1);
    EXPECT_EQ(deadline_ns, 4000000u);
}

TEST_F(QosLoadingTest, TopicFilterRobotCommand) {
    auto qos = provider_->datawriter_qos_w_topic_name(
        "Topics::ProcedureTopics", "RobotCommand");
    auto dur = qos.policy<dds::core::policy::Durability>().kind();
    EXPECT_EQ(dur, dds::core::policy::DurabilityKind::VOLATILE);
}

// --- Participant profiles ---

TEST_F(QosLoadingTest, ParticipantTransport) {
    auto qos = provider_.participant_qos("Participants::Transport");
    (void)qos; // Loading without error is the test
}

TEST_F(QosLoadingTest, ParticipantFactoryDefaults) {
    auto qos = provider_.participant_qos("Factory::FactoryDefaults");
    (void)qos;
}
